sklego/preprocessing/intervalencoder.py

```diff
@@ -0,0 +1,144 @@
+    :return:
+        An array as long as `intervals` that represents the average `y`-values at those intervals,
+        keeping the constraint in mind.
+    """
+    x_internal = np.array([xs >= i for i in intervals]).T.astype(np.float)
+    betas = cp.Variable(x_internal.shape[1])
+    objective = cp.Minimize(cp.sum_squares(x_internal * betas - ys))
+    if method == "increasing":
+        constraints = [betas[i + 1] >= 0 for i in range(betas.shape[0] - 1)]
+    elif method == "decreasing":
+        constraints = [betas[i + 1] <= 0 for i in range(betas.shape[0] - 1)]
+    else:
+        raise ValueError(
+            f"method must be either `increasing` or `decreasing`, got: {method}"
+        )
+    prob = cp.Problem(objective, constraints)
+    prob.solve()
+    return betas.value.cumsum()
+
+
+def _mk_average(xs, ys, intervals, method="average", span=1, **kwargs):
+    """
+    Creates smoothed averages of `ys` at the intervals given by `intervals`.
+    :param xs: all the datapoints of a feature (represents the x-axis)
+    :param ys: all the datapoints what we'd like to predict (represents the y-axis)
+    :param intervals: the intervals at which we'd like to get a good average value
+    :param method: the method that is used for smoothing, can be either `average` or `normal`.
+    :param span: if the method is `average` then this is the span around the interval
+    that is used to determine the average `y`-value, if the method is `normal` the span
+    becomes the value of sigma that is used for weighted averaging
+    :return:
+        An array as long as `intervals` that represents the average `y`-values at those intervals.
+    """
+    results = np.zeros(intervals.shape)
+    for idx, interval in enumerate(intervals):
+        if method == "average":
+            distances = 1 / (0.01 + np.abs(xs - interval))
+            predicate = (xs < (interval + span)) | (xs < (interval - span))
+        elif method == "normal":
+            distances = np.exp(-((xs - interval) ** 2) / span)
+            predicate = xs == xs
+        else:
+            raise ValueError("method needs to be either `average` or `normal`")
+        subset = ys[predicate]
+        dist_subset = distances[predicate]
+        results[idx] = np.average(subset, weights=dist_subset)
+    return results
+
+
+class IntervalEncoder(TransformerMixin, BaseEstimator):
+    """
+    The interval encoder bends features in `X` with regards to`y`.
+    We take each column in X separately and smooth it towards `y` using
+    the strategy that is defined in `method`.
+    Note that this allows us to make certain features strictly monotonic
+    in your machine learning model if you follow this with an appropriate
+    model.
+    :param n_chunks: the number of cuts that makes the interval
+    :param method: the interpolation method used, must be in ["normal", "average"], default: "normal"
```

Comment:
`increasing` and `decreasing` are also valid

Reply:
noted